pkg/controller/kfservice/resources/istio/virtualservice.go

```diff
@@ -0,0 +1,260 @@
+		}
+	}
+
+	// prepare the predict route
+	predictRoute := istiov1alpha3.HTTPRoute{
+		Match: []istiov1alpha3.HTTPMatchRequest{
+			istiov1alpha3.HTTPMatchRequest{
+				URI: &istiov1alpha1.StringMatch{
+					Prefix: fmt.Sprintf("/v1/models/%s:predict", kfsvc.Name),
+				},
+			},
+		},
+		Route: predictRouteDestinations,
+	}
+	httpRoutes = append(httpRoutes, predictRoute)
+
+	// optionally add the explain route
+	explainRouteDestinations := []istiov1alpha3.HTTPRouteDestination{}
+	if kfsvc.Spec.Default.Explainer != nil {
+		explainDefaultSpec, defaultExplainerReason := getExplainStatusConfigurationSpec(&kfsvc.Spec.Default, kfsvc.Status.Default)
+		if explainDefaultSpec != nil {
+			routeDefaultDestination := createHTTPRouteDestination(explainDefaultSpec.Hostname, defaultWeight, r.ingressConfig.IngressServiceName)
+			explainRouteDestinations = append(explainRouteDestinations, routeDefaultDestination)
+
+			explainCanarySpec, canaryExplainerReason := getExplainStatusConfigurationSpec(kfsvc.Spec.Canary, kfsvc.Status.Canary)
+			if explainCanarySpec != nil {
+				routeCanaryDestination := createHTTPRouteDestination(explainCanarySpec.Hostname, canaryWeight, r.ingressConfig.IngressServiceName)
+				explainRouteDestinations = append(explainRouteDestinations, routeCanaryDestination)
+			} else {
+				return nil, createFailedStatus(canaryExplainerReason, "Failed to reconcile canary explainer")
+			}
+		} else {
+			return nil, createFailedStatus(defaultExplainerReason, "Failed to reconcile default explainer")
+		}
+
+		explainRoute := istiov1alpha3.HTTPRoute{
+			Match: []istiov1alpha3.HTTPMatchRequest{
+				istiov1alpha3.HTTPMatchRequest{
+					URI: &istiov1alpha1.StringMatch{
+						Prefix: fmt.Sprintf("/v1/models/%s:explain", kfsvc.Name),
+					},
+				},
+			},
+			Route: predictRouteDestinations,
+		}
+		httpRoutes = append(httpRoutes, explainRoute)
+	}
+
+	vs := istiov1alpha3.VirtualService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        constants.VirtualServiceName(kfsvc.Name),
+			Namespace:   kfsvc.Namespace,
+			Labels:      kfsvc.Labels,
+			Annotations: kfsvc.Annotations,
+		},
+		Spec: istiov1alpha3.VirtualServiceSpec{
+			Hosts: []string{
+				serviceHostname,
+			},
+			Gateways: []string{
+				r.ingressConfig.IngressGateway,
+			},
+			HTTP: httpRoutes,
+		},
+	}
+
+	status := v1alpha2.VirtualServiceStatus{
+		URL:           serviceURL,
+		CanaryWeight:  canaryWeight,
+		DefaultWeight: defaultWeight,
+		Status: duckv1beta1.Status{
+			Conditions: duckv1beta1.Conditions{{
+				Type:   v1alpha2.RoutesReady,
+				Status: corev1.ConditionTrue,
+			}},
+		},
+	}
+
+	return &vs, &status
+}
+
+func extractVirtualServiceHostname(name string, predictorHostName string) string {
+	index := strings.Index(predictorHostName, ".")
+	return name + predictorHostName[index:]
+}
+
+func getPredictStatusConfigurationSpec(endpointSpec *v1alpha2.EndpointSpec, endpointStatusMap *v1alpha2.EndpointStatusMap) (*v1alpha2.StatusConfigurationSpec, string) {
```

Comment:
I found a cleaner interface is be (kfsvc, isCanary). It allows us to have access to the entire object which can be useful when wrapping metadata, etc. Otherwise we will need to keep refactoring this function signature over time (which has happened elsewhere).

Reply:
I see some merit, but would rather not refactor this at this stage... I have opened an issue to add tests for virtualservice.go can revisit at that point?

Reply:
https://github.com/kubeflow/kfserving/issues/417